src/coreComponents/linearAlgebra/interfaces/hypre/HypreSuperlu.cpp

```diff
@@ -0,0 +1,288 @@
+
+namespace geosx
+{
+
+// Check matching requirements on index/value types between GEOSX and SuperLU_Dist
+
+//static_assert( sizeof( int_t ) == sizeof( globalIndex ),
+//               "SuperLU_Dist int_t and geosx::globalIndex must have the same size" );
+
+static_assert( std::is_signed< int_t >::value == std::is_signed< globalIndex >::value,
+               "SuperLU_Dist int_t and geosx::globalIndex must both be signed or unsigned" );
+
+static_assert( std::is_same< double, real64 >::value,
+               "SuperLU_Dist real and geosx::real64 must be the same type" );
+
+namespace
+{
+
+/**
+ * @brief Convert GEOSX globalIndex value to SLUD int_t
+ * @param index the input value
+ * @return the converted value
+ */
+inline int_t toSuperlu_intT( globalIndex const index )
+{
+  return LvArray::integerConversion< int_t >( index );
+}
+
+/**
+ * @brief Converts a non-const array from GEOSX globalIndex type to SLUD int_t
+ * @param[in] index the input array
+ * @return the converted array
+ */
+inline int_t * toSuperlu_intT( globalIndex * const index )
+{
+  return reinterpret_cast< int_t * >( index );
+}
+
+/**
+ * @brief Converts from GEOSX to SuperLU_Dist columns permutation option
+ * @param[in] value the GEOSX option
+ * @return the SuperLU_Dist option
+ */
+colperm_t const & getColPermType( LinearSolverParameters::Direct::ColPerm const & value )
+{
+  static std::map< LinearSolverParameters::Direct::ColPerm, colperm_t > const optionMap =
+  {
+    { LinearSolverParameters::Direct::ColPerm::none, NATURAL },
+    { LinearSolverParameters::Direct::ColPerm::MMD_AtplusA, MMD_AT_PLUS_A },
+    { LinearSolverParameters::Direct::ColPerm::MMD_AtA, MMD_ATA },
+    { LinearSolverParameters::Direct::ColPerm::colAMD, COLAMD },
+    { LinearSolverParameters::Direct::ColPerm::metis, METIS_AT_PLUS_A },
+    { LinearSolverParameters::Direct::ColPerm::parmetis, PARMETIS },
+  };
+
+  GEOSX_LAI_ASSERT_MSG( optionMap.count( value ) > 0, "Unsupported SuperLU_Dist columns permutation option: " << value );
+  return optionMap.at( value );
+}
+
+/**
+ * @brief Converts from GEOSX to SuperLU_Dist rows permutation option
+ * @param[in] value the GEOSX option
+ * @return the SuperLU_Dist option
+ */
+rowperm_t const & getRowPermType( LinearSolverParameters::Direct::RowPerm const & value )
+{
+  static std::map< LinearSolverParameters::Direct::RowPerm, rowperm_t > const optionMap =
+  {
+    { LinearSolverParameters::Direct::RowPerm::none, NOROWPERM },
+    { LinearSolverParameters::Direct::RowPerm::mc64, LargeDiag_MC64 },
+  };
+
+  GEOSX_LAI_ASSERT_MSG( optionMap.count( value ) > 0, "Unsupported SuperLU_Dist rows permutation option: " << value );
+  return optionMap.at( value );
+}
+
+/**
+ * @brief Converts a matrix from Hypre to SuperLU_Dist format
+ * @param[in] matrix the HypreMatrix object
+ * @param[out] SLUDData the structure containing the matrix in SuperLU_Dist format
+ */
+void ConvertToSuperMatrix( HypreMatrix const & matrix,
```

Comment:
I don't understand??

Reply:
I'm not familiar with how `HypreMatrix` works. But since it can now contain device pointers is there any chance the following code will pull out a device pointer?

Reply:
Certainly. But we can't. really call SuperLU from hypre while running on the GPU until we setup SuperLU to run on the GPU.